app/stac_api/views_mixins.py

```diff
@@ -57,28 +59,46 @@ class UpdateModelMixin:
     def get_write_request_data(self, request, *args, partial=False, **kwargs):
         return request.data
 
-    def update(self, request, *args, **kwargs):
+    @transaction.atomic
+    def upsert(self, request, *args, **kwargs):
         partial = kwargs.pop('partial', False)
         serializer_kwargs = {'partial': partial}
-        instance = self.get_object()
         data = self.get_write_request_data(request, partial=partial, *args, **kwargs)
-        serializer = self.get_serializer(instance, data=data, **serializer_kwargs)
-        serializer.is_valid(raise_exception=True)
-        self.perform_update(serializer)
 
-        if getattr(instance, '_prefetched_objects_cache', None):
-            # If 'prefetch_related' has been applied to a queryset, we need to
-            # forcibly invalidate the prefetch cache on the instance.
-            instance._prefetched_objects_cache = {}  # pylint: disable=protected-access
+        # update
+        if self.check_existence():
+            instance = self.get_object(called_from_within_atomic=True)
+            serializer = self.get_serializer(instance, data=data, **serializer_kwargs)
+            serializer.is_valid(raise_exception=True)
+            self.perform_upsert(serializer)
+
+            if getattr(instance, '_prefetched_objects_cache', None):
+                # If 'prefetch_related' has been applied to a queryset, we need to
+                # forcibly invalidate the prefetch cache on the instance.
+                instance._prefetched_objects_cache = {}  # pylint: disable=protected-access
 
-        return Response(serializer.data)
+            return Response(serializer.data)
+
+        # else create
+        serializer = self.get_serializer(data=data)
+        serializer.is_valid(raise_exception=True)
+        self.perform_upsert(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
 
-    def perform_update(self, serializer):
+    def perform_upsert(self, serializer):
```

Comment:
What's the reason for putting this one call into a dedicated function?

Reply:
The reason is to be consistent with the way, it currently is implemented. Currently, there's the `perform_update()`, which I simply renamed to `perform_upsert()`. And the currently implemented way is simply an overriden original mixin. So the `perform_XYZ()` comes from the original Django mixins.